Make it clear that code outside plugins/ can have other free software licenses as well (#212)
ansible-collections/overview: collection_requirements.rst

               <https://opensource.org/licenses/BSD-2-Clause>`_ license to make it possible for
               third-party modules which are licensed incompatibly with the GPLv3 to use them.
               Please consider this use case when licensing your own ``module_utils``.
:All other code in ``plugins/``: All other code in ``plugins/`` must be under the `GPL-3.0-or-later
                                 <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_.  These plugins
                                 are run inside of the Ansible controller process which is licensed under
                                 the GPLv3+ and often must import code from the controller.  For these
                                 reasons, the GPLv3+ must be used.

> :All other code: Code outside ``plugins/``, in particular unit tests, must be licensed with

Reviewer: Sorry for not batching up my feedback so you could apply it all at once.

```suggestion
:All other code: Code outside ``plugins/``, in particular unit tests for ``modules`` and ``module_utils``, must be licensed with
```

Author: No problem about batching :)

I'm not sure why unit tests for modules and module_utils should be singled out here. Other unit tests also have to be licensed with a free software license (though very likely with GPLv3+, at least the way I interpret it, since IANAL).

Reviewer: Well, wouldn't the tests for the other plugins import the code from the other plugins that have to be GPL-licensed?

Author: Yes, but (there's always a but ;) ) you can also have Python files there that do not import non-module/module_utils plugins. For example you could have a Apache 2.0 licensed file that provides some utility that is only used by a test for an inventory plugin, and thus located right next to it. Or you can have generic unit test helpers in `tests/unit/` outside of `tests/unit/plugins/modules/` and `tests/unit/plugins/module_utils/` (like the mock import helper) that are licensed under another license than GPLv3+.

Reviewer: Maybe it makes sense to phrase it like so?

```
:All other code: Code outside ``plugins/``, in particular unit tests [for ``modules`` and ``module_utils``], may be licensed under
                 another free software license that is compatible with the
                 `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_,
                 provided that such code that does not import any code that is licensed under
                 the ``GPL-3.0-or-later``.
```

(The bracketed part would either be removed or included without the brackets)

I don't feel strongly either way, but I think my suggestion might be clearer.

Author: Hmm, I think this is only easier to understand for folks who know that code importing GPLv3+ licensed code needs to be GPLv3+ licensed as well. How about adding something like "Code satisfying the latter must be licensed under the `GPL-3.0-or-later`."?

Reviewer: I've edited the suggestion.

Author: Applied as 0a1b0f352485e1ddbbac944130499b2b2a205440 (without the square brackets).

               <https://opensource.org/licenses/BSD-2-Clause>`_ license to make it possible for
               third-party modules which are licensed incompatibly with the GPLv3 to use them.
               Please consider this use case when licensing your own ``module_utils``.
:All other code in ``plugins/``: All other code in ``plugins/`` must be under the `GPL-3.0-or-later
                                 <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_.  These plugins
                                 are run inside of the Ansible controller process which is licensed under

>                                  the GPLv3+ and often must import code from the controller.  For these

Reviewer: ```suggestion
                                 the ``GPL-3.0-or-later`` and often must import code from the controller.  For these
```

                                 reasons, the GPLv3+ must be used.
:All other code: Code outside ``plugins/``, in particular unit tests, may be licensed under
                 another free software license that is compatible with the
                 `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_,

>                  provided that such code that does not import any other code that is licensed under

Reviewer 2: ```suggestion
                 provided that such code  does not import any other code that is licensed under
```

Reviewer 3: Since unit tests import modules, plugins or ansible-core code, it's probably worth clarifying that here.

               <https://opensource.org/licenses/BSD-2-Clause>`_ license to make it possible for
               third-party modules which are licensed incompatibly with the GPLv3 to use them.
               Please consider this use case when licensing your own ``module_utils``.
:All other code in ``plugins/``: All other code in ``plugins/`` must be under the `GPL-3.0-or-later
                                 <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_.  These plugins
                                 are run inside of the Ansible controller process which is licensed under
                                 the ``GPL-3.0-or-later`` and often must import code from the controller.

>                                  For these reasons, the GPLv3+ must be used.

Reviewer: `s/GPLv3+/GPL-3.0-or-later/`

Author: Implemented in 48da08635ce9bdf7a571572610f17f3dbaae446b.

               <https://opensource.org/licenses/BSD-2-Clause>`_ license to make it possible for
               third-party modules which are licensed incompatibly with the GPLv3 to use them.
               Please consider this use case when licensing your own ``module_utils``.
:All other code in ``plugins/``: All other code in ``plugins/`` must be under the `GPL-3.0-or-later
                                 <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_.  These plugins
                                 are run inside of the Ansible controller process which is licensed under
                                 the ``GPL-3.0-or-later`` and often must import code from the controller.
                                 For these reasons, the GPLv3+ must be used.

> :All other code: Code outside ``plugins/``, in particular unit tests, may be licensed under

Reviewer 3: This section is long enough that I'm concerned someone will see the "in particular unit tests" part and not notice the caveat at the end regarding unit tests. Perhaps it would be adequate to remove the "in particular unit tests" part of the sentence?

Reviewer 3: Or perhaps unit tests should have their own section ahead of "all other code" so we can more clearly address the topic there without making this section longer.

Author: How about rewriting this section as
```
:All other code: Code outside ``plugins/`` may be licensed under another free software license that is compatible
                 with the `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_,
                 provided that such code does not import any other code that is licensed under
                 the ``GPL-3.0-or-later``. If the file does import other ``GPL-3.0-or-later`` code,
                 then it must similarly be licensed under ``GPL-3.0-or-later``. Note that this applies in
                 particular to unit tests; these often import code from ansible-core, plugins, module utils,
                 or modules, and such code is often licensed under ``GPL-3.0-or-later``.
```

Reviewer 3: I think that would be adequate.

Author: I've added it in 626b67e21400d9670308e99b4535277662025eb7. Thanks!

               <https://opensource.org/licenses/BSD-2-Clause>`_ license to make it possible for
               third-party modules which are licensed incompatibly with the GPLv3 to use them.
               Please consider this use case when licensing your own ``module_utils``.
:Code outside plugins/: Code outside `plugins/`, in particular unit tests, must be licensed with
                        a free software license that is compatible with the
                       `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_.
                       Note that such code that imports GPL licensed code must be GPL licensed

>                        itself as well.

Reviewer: This does not seem to be formatted correctly. The indentation is really wonky in the rendered version.

Author: Fixed in 964631699a266c6ccd3e4348b2d164a2a3a91e45.
